### PR TITLE
Use types and typeid properties consistent with HOOMD

### LINKED ISSUE #137: Use types and typeid properties consistent with HOOMD
## Description

The properties for type names and ids are inconsistent with HOOMD, and within the Trajectory / Frame APIs.

|  API     | HOOMD/GSD                                     | garnett Frame                                       | garnett Trajectory                                                     |
|----------|-----------------------------------------------|-----------------------------------------------------|------------------------------------------------------------------------|
| type     | N/A                                           | N/A                                                 | array of N_types strings, ['A', 'B', 'C']                              |
| types    | array of N_types  strings, ['A', 'B', 'C']    | array of N_particles  strings, ['A', 'A', 'B', ...] | array of (N_frames, N_particles)  strings, [['A', 'A', 'B', ...], ...] |
| typeid   | array of N_particles indices,  [0, 0, 1, ...] | N/A                                                 | N/A                                                                    |
| type_ids | N/A                                           | N/A                                                 |                                                                        |

## Proposed solution

Refactor in reader classes:
- The type id (unsigned integer) should be the "fundamental" identifier, not the type name string

Refactor in Trajectory:
- `types` should be removed
- `type` should be removed
- `types` should be a list of lists of strings with unique types for each frame (N_frames, N_types)
- `type_ids` should become `typeid`

Refactor in Frame:
- `types` should be a list of type name strings
- `typeid` should be an array of integer type indices

## Additional context

N/A

### PR DESCRIPTION
**Pull request by bdice**
## Description
This PR rewrites the internals for `types` and `typeid` to match HOOMD's conventions.

## Motivation and Context
Resolves: #137.

## How Has This Been Tested?
The meanings of the type-related properties were changed, so I changed the expected behavior of the tests to match.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds or improves functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/garnett/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation (if relevant).
- [x] I have added tests that cover my changes (if relevant).
- [x] All new and existing tests passed.
- [x] I have updated the [credits](https://github.com/glotzerlab/garnett/blob/master/doc/credits.rst).
- [x] I have updated the [Changelog](https://github.com/glotzerlab/garnett/blob/master/changelog.rst).

### CODE REVIEW

File: garnett/dcdfilereader.py [hunk 1]
```diff
@@ -180,17 +184,18 @@ def read(self):
...
-                raw_frame.shapedef = copy.deepcopy(self.t_frame.shapedef)
+                raw_frame.type_shapes = copy.deepcopy(self.t_frame.type_shapes)
             except AttributeError:
                 pass
         if not self._loaded():
             self._load()
         assert self._loaded()
         raw_frame.box = self._box
-        raw_frame.types = copy.copy(self._types)
+        raw_frame.types = self._types
```
Comment (bdice) [inline garnett/dcdfilereader.py]: It's not clear to me why the `copy.copy` was required before. It may not be needed?

File: garnett/dcdfilereader.py [hunk 2]
```diff
@@ -153,9 +154,11 @@ def _load(self, xyz=None, ort=None):
             ort = np.zeros((N, 4), dtype=self._dtype)
         self._read(xyz=xyz)
         if self.t_frame is None:
-            self._types = [self.default_type] * len(self)
+            self._types = [self.default_type]
+            self._typeid = np.zeros(N, dtype=np.uint)
         else:
             self._types = self.t_frame.types
```
Comment (bdice) [inline garnett/dcdfilereader.py]: Maybe we need to `copy.copy` the fields that come from the template frame?

**Comment by bdice** [comment]
@glotzerlab/garnett-maintainers This should be ready for review.

The `types` and `typeid` properties are now a consistent source of truth for type information.

`types` looks like `['A', 'B', 'C']`. Strings with each type name, of length T (number of types, type `str`).
`typeid` looks like `[0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2]`. It's an array of indices of length N (number of particles, type `np.uint`) corresponding to each particle's type.

Because shape definitions are closely linked to our handling of types, I also re-wrote support for `type_shapes` and `shapedef` in this PR. The primary source of truth for shapes is now `type_shapes`. The `shapedef` attribute (which is an `OrderedDict`) was retained for compatibility but it's not fully functional. Setting shape definitions with `frame.shapedef['A'] = sphere_object` won't work, but setting shape definitions with `frame.shapedef = {'A': sphere_object}` will work. I am not sure how to improve on this without significantly complicating that code. I think we should deprecate/remove `shapedef` and encourage the use of `type_shapes` in its place.

Tests were only edited for two reasons:
- The meanings of the type-related properties were changed, so I changed the expected behavior of the tests to match.
- `flake8` is now applied to tests as well on CI.

Note: We were previously running flake8 only on the `garnett/` directory on CI, but the tests directory was not excluded in the `.flake8` config file. That meant that `flake8` would fail locally and pass on CI, which isn't ideal. I did not want to separate these changes since there were not many changes needed and I was offline while doing this development work (it's slightly harder to manage PRs/branches while offline). If you feel strongly about that, I can revert that and separate it into a new PR.

File: doc/quickstart.rst [hunk 3]
```diff
@@ -92,12 +93,11 @@ Supported properties are listed below:
     traj.moment_inertia  # MxNx3
     traj.angmom          # MxNx4
     traj.image           # MxNx4
-    traj.types           # MxN
     traj.type_ids        # MxN
-    traj.type            # list of type names ordered by type_id
 
-    # where M=len(traj) is the number of frames and N=max((len(f) for f in traj))
-    # is the is the maximum number of particles in any frame.
+    # where M is the number of frames,
```
Comment (lyrivera) [inline doc/quickstart.rst]: Fix `type_ids` in the list above.

File: changelog.rst [hunk 4]
```diff
@@ -18,6 +18,7 @@ Added
 Changed
 +++++++
   - Updated GSD reader to use the GSD v2.0.0 API.
+  - Changed behavior of ``types``, ``typeid``, ``type_shapes`` to match HOOMD conventions.
```
Comment (lyrivera) [inline changelog.rst]: Add deprecation note for the `shapedef` property. Maybe also mention how it works from now on until its removal.

File: README.md [hunk 5]
```diff
@@ -72,16 +77,19 @@ traj.charge          # MxN
 traj.diameter        # MxN
 traj.moment_inertia  # MxNx3
 traj.angmom          # MxNx4
-traj.types           # MxN
+traj.types           # MxT
```
Comment (lyrivera) [inline README.md]: Add `traj.N` and `traj.typeid` to the list.

**Review by lyrivera — COMMENTED** [review]
@bdice Thanks for this work. Left a few comments below. @klywang This is a big refactor so it would be nice if you could also provide a review.

File: garnett/gsdhoomdfilewriter.py [hunk 6]
```diff
@@ -10,74 +10,12 @@
...
-from .shapes import SphereShape, ConvexPolyhedronShape, ConvexSpheropolyhedronShape, \
-    PolygonShape, SpheropolygonShape, EllipsoidShape
-from .errors import GSDShapeError
 
 from .trajectory import PARTICLE_PROPERTIES
 
 logger = logging.getLogger(__name__)
 
 
-def _write_shape_definitions(snap, shapedefs):
```
[the remark this answers is not in the document]
Comment (bdice) [reply garnett/gsdhoomdfilewriter.py]: Yes, that was dropped in favor of writing `type_shapes` in #127. I mentioned it in a note on that PR description. Otherwise we could be writing HPMC state into a DEM simulation, for example. I added a specific note in the changelog.

File: garnett/posfilewriter.py [hunk 7]
```diff
@@ -95,10 +95,8 @@ def _write(msg, end='\n'):
 
             # shape defs
             try:
-                required = set(frame.types).intersection(
-                    set(frame.shapedef.keys()))
-                not_defined = set(frame.types).difference(
-                    set(frame.shapedef.keys()))
+                required = [t for t in frame.types if t in frame.shapedef]
```
[the remark this answers is not in the document]
Comment (bdice) [reply garnett/posfilewriter.py]: This rewritten code may be less robust in a few edge cases where there are undefined types, or more type names than type_shapes, etc. It handles those cases less well because the specification with `type_shapes` is a little more rigorously defined. I think we can accept it as-is if tests pass, and handle those edge cases if users need them. I would suspect those would only be seen in very old pre-HOOMD POS files.

File: garnett/trajectory.py [hunk 8]
```diff
@@ -259,12 +280,14 @@ def _validate_input_array(self, value, dim, nelem=None, dtype=None):
...
         """Generate a frame object from a raw frame object."""
-        N = len(raw_frame.types)
+        N = len(raw_frame.position)
         ret = FrameData()
 
         mapping = dict()
-        for prop in PARTICLE_PROPERTIES:
-            mapping[prop] = np.asarray(getattr(raw_frame, prop), dtype=dtype)
+        for prop, dtype_ in PARTICLE_PROPERTIES.items():
+            if dtype_ == DEFAULT_DTYPE:
```
[the remark this answers is not in the document]
Comment (bdice) [reply garnett/trajectory.py]: My solution is less than ideal. We can handle `float32` or `float64` in garnett, and I'm just trying to detect that. I don't think adding a check for `dtype is not None` is helpful because all of the data types are defined at the top of this file. A `None` datatype would concern me here, but I don't think it's possible to have that.

[the remark this answers is not in the document]
Comment (lyrivera) [reply garnett/trajectory.py]: I only mentioned it because the `dtype=None` is de default argument for the function. However, now I see that  the function is called only once and a proper argument is always passed. Perhaps removing the default `None` would avoid confusion in the future.

File: garnett/gsdhoomdfilewriter.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lyrivera) [reply garnett/gsdhoomdfilewriter.py]: Thanks for the reminder. It totally slipped my mind.

File: garnett/posfilewriter.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (lyrivera) [reply garnett/posfilewriter.py]: What I meant to ask was why were you using the `shapedef` property instead of the `type_shapes`, since `shapedef` will be removed in the future. But I can see your point now.

**Comment by bdice** [comment]
@lyrivera I'm not sure what's left to do on this -- if there is further work to be done, could I do it in a follow-up PR? There are a few PRs waiting on this so we shouldn't delay it longer than necessary.

**Comment by lyrivera** [comment]
> @lyrivera I'm not sure what's left to do on this -- if there is further work to be done, could I do it in a follow-up PR? There are a few PRs waiting on this so we shouldn't delay it longer than necessary.

@bdice I saw you marked most of the comments I left as resolved but I don't see any further changes. Did you push your latest work? If not, you can certainly address them in a follow-up PR.

**Comment by bdice** [comment]
@lyrivera Oh. Right. I didn't push. 😅 I can't find which computer I made these changes on at the moment. I'll follow up once the changes are pushed.

**Comment by lyrivera** [comment]
> @lyrivera Oh. Right. I didn't push. 😅 I can't find which computer I made these changes on at the moment. I'll follow up once the changes are pushed.

I knew something was wrong lol. No worries just request my review again once you push.

File: garnett/trajectory.py [hunk 9]
```diff
@@ -496,15 +525,37 @@ def box(self, value):
...
+    @property
+    def typeid(self):
+        "N array of type indices for N particles."
+        self.load()
+        return self.frame_data.typeid
+
+    @typeid.setter
+    def typeid(self, value):
+        self.load()
+        self.frame_data.typeid = value
```
[the remark this answers is not in the document]
Comment (bdice) [reply garnett/trajectory.py]: I applied this change to `typeid`. I'm not sure if we can use the same validation logic for `type_shapes` and `types`, since they're non-numeric.

[the remark this answers is not in the document]
Comment (lyrivera) [reply garnett/trajectory.py]: Makes sense. Perhaps in the future we can slightly modify the validation to check non-numeric arrays but it's not critical now.

**Review by lyrivera — COMMENTED** [review]
Left one more comment to consider. Once its addressed I will approve. Thanks!

**Comment by bdice** [comment]
@lyrivera I think this is probably ready.

File: garnett/dcdfilereader.py [hunk 10]
```diff
@@ -233,17 +237,15 @@ def load_arrays(self):
             if not frame._loaded():
                 frame._load(xyz=xyz[i], ort=ort[i])
 
-        # Types, Can only be handled after frame._load() calls.
-        types = [f._types for f in self.frames]
-        type_ids = np.zeros((len(self), N), dtype=np.int_)
-        _type = _generate_type_id_array(types, type_ids)
+        # Types can only be handled after frame._load() calls.
```
[the remark this answers is not in the document]
Comment (bdice) [reply garnett/dcdfilereader.py]: I fixed the dtype. @klarh says it looks like that info is encoded once at the top of the file, not per-frame. I'm fine with making the constant-N assumption and fixing it later, if it's ever needed.

[the remark this answers is not in the document]
Comment (lyrivera) [reply garnett/dcdfilereader.py]: Sounds good.

File: garnett/posfilewriter.py [hunk 11]
```diff
@@ -95,17 +95,10 @@ def _write(msg, end='\n'):
...
-                not_defined = set(frame.types).difference(
-                    set(frame.shapedef.keys()))
-                for name in required:
-                    _write('def {} "{}"'.format(name, frame.shapedef[name].pos_string))
-                for name in not_defined:
-                    logger.info(
-                        "No shape defined for '{}'. "
-                        "Using fallback definition.".format(name))
-                    _write('def {} "{}"'.format(name, DEFAULT_SHAPE_DEFINITION.pos_string))
+                if len(frame.types) != len(frame.type_shapes):
```
Comment (lyrivera) [inline garnett/posfilewriter.py]: I'm OK with erroring out if there are not enough shape definitions but it should probably be mentioned in the `changelog` since its a change from the expected behavior.

**Review by lyrivera — APPROVED** [review]
Left one last minor comment, but other than that, looks good to go! Thanks again!
